README.md
# Fitted TextField Container

A `Container` that automatically resizes it's width to fit the containing `TextField`'s text value.

Read the short blog: https://medium.com/@huyffs/fitted-textfield-container-for-flutter-c9c63644441c

## Usage

### FittedTextFieldContainer

| Param        | Type                                                     | Default                                | Description                                                                                                                                                               |
|--------------|----------------------------------------------------------|----------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `child`      | `TextField`                                              | * _required_                           | The `TextField` to fit                                                                                                                                                    |
| `calculator` | `CalculateFunction`                                      | `FittedTextFieldCalculator.fitVisible` | A function to calculate the width. The default function fits the visible content & decorations (i.e. it collapses to the content/label's width when hint text disappears) |
| `builder`    | `Widget Function(BuildContext context, TextField child)` | `null`                                 | A builder that can be used to build complex layouts. The `child` TextField is provided to the build function                                                              |

#### Example

```dart
FittedTextFieldContainer(
    child: TextField(
        controller: someTextEditingController,
        textAlign: TextAlign.right,
        decoration: InputDecoration(
            prefixText: "£",
            labelText: "Amount",
        ),
    ),
)
```

### AnimatedFittedTextFieldContainer
All params from `FittedTextFieldContainer` plus the following

| Param            | Type       | Default                       | Description                                                        |
|------------------|------------|-------------------------------|--------------------------------------------------------------------|
| `growCurve`      | `Curve`    | `Curves.easeOutCirc`          | The curve to use in the grow animation                             |
| `growDuration`   | `Duration` | `Duration(milliseconds: 300)` | Duration to animate the container's width when `TextField` grows   |
| `shrinkCurve`    | `Curve`    | `Curves.easeInCirc`           | The curve to use in the shrink animation                           |
| `shrinkDuration` | `Duration` | `Duration(milliseconds: 600)` | Duration to animate the container's width when `TextField` shrinks |

#### Example

```dart
AnimatedFittedTextFieldContainer(
    growDuration: Duration(milliseconds: 300),
    shrinkDuration: Duration(milliseconds: 600),
    growCurve: Curves.easeOutCirc,
    shrinkCurve: Curves.easeInCirc,
    child: TextField(
        controller: someTextEditingController,
        textAlign: TextAlign.right,
        decoration: InputDecoration(
            prefixText: "£",
            labelText: "Amount",
        ),
    ),
)
```

### `FittedTextFieldMeasurer`
A class that takes a `TextField` and `TextStyle` and returns the following measurements:
* `textField` The fitted textfield
* `textStyle` The fitted textfield's style or the default style
* `textWidth` Width of the text content
* `labelWidth` Width of the label
* `hintWidth` Width of the hint text
* `fixedWidths` Width of fixed elements (prefixText, suffixText, contentPadding and cursorWidth)

### `CalculateFunction`
This function takes measured values and returns a width value. 

#### Signature:
```dart
double CalculateFunction(FittedTextFieldMeasurer)
```

#### Example:
```dart
FittedTextFieldContainer(
    calculator: (m) =>
        m.fixedWidths +
        max(m.labelWidth, max(m.hintWidth, m.textWidth)),
    child: TextField(
        controller: someTextEditingController,
        textAlign: TextAlign.right,
        decoration: InputDecoration(
            hintText: "Width of hint text",
        ),
    ),
);
```
#### Predefined functions
A few calculators are provided for common use-cases
* `fitAll` fits content, label and hint texts
* `fitVisible` fits the visible content & decorations (i.e. it collapses to the content/label's width when hint text disappears)
* `fitVisibleWithPadding(double padding)` in addition to fitting visible content it adds a fixed padding value
* `fitVisibleWithRange(double min, double max)` in addition to fitting visible content it enforces a minimum and (optional) maximum width

#### Example:
Fit all
```dart
FittedTextFieldContainer(
    calculator: FittedTextFieldCalculator.fitAll,
    child: ...
);
```
Fit visible with minimum
```dart
FittedTextFieldContainer(
    calculator: FittedTextFieldCalculator.fitVisibleWithRange(120),
    child: ...
);
```
Fit visible with maximum
```dart
FittedTextFieldContainer(
    calculator: FittedTextFieldCalculator.fitVisibleWithRange(0, 180),
    child: ...
);
```

### Using the `builder`
```dart
AnimatedFittedTextFieldContainer(
    child: TextField(
        controller: someTextEditingController,
        textAlign: TextAlign.right,
        decoration: InputDecoration(
            labelText: "Twinkle twinkle",
        ),
    ),
    builder: (context, child) => Container(
        child: Stack(
            overflow: Overflow.visible,
            children: <Widget>[
                child,
                Positioned(
                    top: 0,
                    right: -22,
                    child: _aniWholePoundCtl.text.isEmpty
                        ? Icon(Icons.star_border)
                        : Icon(Icons.star, color: Colors.amber),
                ),
            ],
        ),
    ),
)
```

## Contribute

Pull requests are welcome!

## Licence

MIT License

Copyright (c) 2020 Huy Dinh

CHANGELOG.md
## 2.0.0
* Null safety - thanks to [AliAmmar](https://github.com/Ali1Ammar)

## 1.3.1+1
* Fixed README

## 1.3.1
* Updated README
* Added width calculators for common use-cases

## 1.3.0
* Added `calculator` property to allow custom width calculations 
* **Breaking Change.** Removed `minWidth`, `maxWidth`, `prefixIconWidth`, `suffixIconWidth` properties (they can be replaced with the custom width calculator)
* **Breaking Change.** Defaults to collapse width to visible text instead of keeping the hint text (Thanks [AdamJonsson](https://github.com/AdamJonsson) for PR [#2](https://github.com/huyffs/fitted_text_field_container/pull/2))
* Include padding into calculations (Thanks [David King](https://github.com/oodavid) for PR [#3](https://github.com/huyffs/fitted_text_field_container/pull/3))

## 1.2.1
* Fixed exception thrown when `textDirection` is not set
* Fixed incorect calculations due to default stules not being available durint initialisation

## 1.2.0
* Added `builder` for building complex layouts
* Fixed calculations to remove the need for a magic number
* Code tidy-ups and optimisations which should improve performance by a tiny tiny bit
* Fixed docs and typos (Thanks [jpnurmi](https://github.com/jpnurmi) for PR [#1](https://github.com/huyffs/fitted_text_field_container/pull/1))

## 1.1.0
* Added `prefixIconWidth` for specifying the width of the prefix icon (if used)
* Added `suffixIconWidth` for specifying the width of the suffix icon (if used)
* Added `minWidth` property for specifying a minimum width
* Added `maxWidth` property for specifying a maximum width
* Minimum width takes `labelText` and `hintText` into account (if specified) 

## 1.0.0+3
* Updated demo & documentation

## 1.0.0+2
* Formatted code

## 1.0.0+1
* Initial release - take 2

## 1.0.0
* Initial release

lib/src/utils.dart
import 'package:flutter/material.dart';

class TextFieldMeasurer {
  TextFieldMeasurer._({
    required this.textField,
    required this.textStyle,
    required this.hintWidth,
    required this.labelWidth,
    required this.fixedWidths,
  });

  factory TextFieldMeasurer.create(TextField textField, TextStyle? textStyle) {
    final prefixWidth = _getPrefixTextSize(textField, textStyle).width;
    final suffixWidth = _getSuffixTextSize(textField, textStyle).width;
    final hintWidth = _getHintTextSize(textField, textStyle).width;
    final labelWidth = _getLabelTextSize(textField, textStyle).width;
    final paddingWidth = textField.decoration!.contentPadding != null
        ? textField.decoration!.contentPadding!.collapsedSize.width
        : 0;
    // Add enough space for the cursor to prevent it being positined onto the
    // next line in a multiline textfield and scrolled in a single-line text
    // field
    final cursorWidth = textField.cursorWidth + 1;

    return TextFieldMeasurer._(
      textField: textField,
      textStyle: textStyle,
      hintWidth: hintWidth,
      labelWidth: labelWidth,
      fixedWidths: prefixWidth + suffixWidth + paddingWidth + cursorWidth,
    );
  }

  final TextField textField;
  final TextStyle? textStyle;
  final double hintWidth;
  final double labelWidth;
  final double fixedWidths;

  static Size _calculateSize({
    required String? text,
    required TextStyle? style,
    required TextAlign textAlign,
    required StrutStyle? strutStyle,
    required int? maxLines,
    required TextDirection textDirection,
  }) {
    if (text == null || text.isEmpty) {
      return Size(0, 0);
    }
    final painter = TextPainter(
      text: TextSpan(
        style: style,
        text: text,
      ),
      textAlign: textAlign,
      strutStyle: strutStyle,
      maxLines: maxLines,
      textDirection: textDirection,
    )..layout();
    return painter.size;
  }

  static Size _getPrefixTextSize(TextField textField,
          [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.prefixText,
        style: textField.decoration!.prefixStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  static Size _getSuffixTextSize(TextField textField,
          [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.suffixText,
        style: textField.decoration!.suffixStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  static Size _getHintTextSize(TextField textField, [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.hintText,
        style: textField.decoration!.hintStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  static Size _getLabelTextSize(TextField textField,
          [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.labelText,
        style: textField.decoration!.labelStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  double get textWidth => textField.controller!.text.isNotEmpty
      ? _calculateSize(
          text: textField.controller!.text,
          style: textField.style ?? textStyle,
          textAlign: textField.textAlign,
          strutStyle: textField.strutStyle,
          maxLines: textField.maxLines,
          textDirection: textField.textDirection ?? TextDirection.ltr,
        ).width
      : 0;
}


lib/src/measurer.dart
import 'package:flutter/material.dart';

class FittedTextFieldMeasurer {
  FittedTextFieldMeasurer._({
    required this.textField,
    required this.textStyle,
    required this.hintWidth,
    required this.labelWidth,
    required this.fixedWidths,
  });

  factory FittedTextFieldMeasurer.create(
      TextField textField, TextStyle? textStyle) {
    final prefixWidth = _getPrefixTextSize(textField, textStyle).width;
    final suffixWidth = _getSuffixTextSize(textField, textStyle).width;
    final hintWidth = _getHintTextSize(textField, textStyle).width;
    final labelWidth = _getLabelTextSize(textField, textStyle).width;
    final paddingWidth = textField.decoration!.contentPadding != null
        ? textField.decoration!.contentPadding!.collapsedSize.width
        : 0;
    // Add enough space for the cursor to prevent it being positined onto the
    // next line in a multiline textfield and scrolled in a single-line text
    // field
    final cursorWidth = textField.cursorWidth + 1;

    return FittedTextFieldMeasurer._(
      textField: textField,
      textStyle: textStyle,
      hintWidth: hintWidth,
      labelWidth: labelWidth,
      fixedWidths: prefixWidth + suffixWidth + paddingWidth + cursorWidth,
    );
  }

  final TextField textField;
  final TextStyle? textStyle;
  final double hintWidth;
  final double labelWidth;
  final double fixedWidths;

  static Size _calculateSize({
    required String? text,
    required TextStyle? style,
    required TextAlign textAlign,
    required StrutStyle? strutStyle,
    required int? maxLines,
    required TextDirection textDirection,
  }) {
    if (text == null || text.isEmpty) {
      return Size(0, 0);
    }
    final painter = TextPainter(
      text: TextSpan(
        style: style,
        text: text,
      ),
      textAlign: textAlign,
      strutStyle: strutStyle,
      maxLines: maxLines,
      textDirection: textDirection,
    )..layout();
    return painter.size;
  }

  static Size _getPrefixTextSize(TextField textField,
          [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.prefixText,
        style: textField.decoration!.prefixStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  static Size _getSuffixTextSize(TextField textField,
          [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.suffixText,
        style: textField.decoration!.suffixStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  static Size _getHintTextSize(TextField textField, [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.hintText,
        style: textField.decoration!.hintStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  static Size _getLabelTextSize(TextField textField,
          [TextStyle? defaultStyle]) =>
      _calculateSize(
        text: textField.decoration!.labelText,
        style: textField.decoration!.labelStyle ?? defaultStyle,
        textAlign: textField.textAlign,
        strutStyle: textField.strutStyle,
        maxLines: textField.maxLines,
        textDirection: textField.textDirection ?? TextDirection.ltr,
      );

  double get textWidth => textField.controller!.text.isNotEmpty
      ? _calculateSize(
          text: textField.controller!.text,
          style: textField.style ?? textStyle,
          textAlign: textField.textAlign,
          strutStyle: textField.strutStyle,
          maxLines: textField.maxLines,
          textDirection: textField.textDirection ?? TextDirection.ltr,
        ).width
      : 0;
}


lib/src/calculator.dart
import 'dart:math' show max;

import 'package:fitted_text_field_container/src/measurer.dart';

typedef double CalculateFunction(FittedTextFieldMeasurer measurer);

class FittedTextFieldCalculator {
  /// FittedTextFieldCalculator contains calculators for common use-cases

  /// fitAll fits content, label and hint texts
  static double fitAll(FittedTextFieldMeasurer m) =>
      m.fixedWidths +
      max<double>(m.labelWidth, max<double>(m.hintWidth, m.textWidth));

  /// fitVisible fits the visible content & decorations (i.e. it collapses to
  /// the content/label's width when hint text disappears)
  static double fitVisible(FittedTextFieldMeasurer m) =>
      m.fixedWidths +
      max<double>(m.labelWidth,
          m.textField.controller!.text.isEmpty ? m.hintWidth : m.textWidth);

  /// fitVisibleWithPadding in addition to fitting visible content it adds a
  /// fixed padding value
  static CalculateFunction fitVisibleWithPadding(double padding) =>
      (FittedTextFieldMeasurer m) =>
          padding +
          m.fixedWidths +
          max<double>(m.labelWidth,
              m.textField.controller!.text.isEmpty ? m.hintWidth : m.textWidth);

  /// fitVisibleWithRange in addition to fitting visible content it enforces
  /// a minimum and (optional) maximum width
  static CalculateFunction fitVisibleWithRange(double minWidth,
          [double? maxWidth]) =>
      (FittedTextFieldMeasurer? m) {
        final width = m!.fixedWidths +
            max<double>(
                m.labelWidth,
                m.textField.controller!.text.isEmpty
                    ? m.hintWidth
                    : m.textWidth);
        if (width < minWidth) {
          return minWidth;
        } else if (maxWidth != null && width > maxWidth) {
          return maxWidth;
        } else {
          return width;
        }
      };
}


lib/src/fitted_text_field_container.dart
import 'package:fitted_text_field_container/src/calculator.dart';
import 'package:flutter/material.dart';

import 'measurer.dart';

class FittedTextFieldContainer extends StatefulWidget {
  /// Creates an animated container to wrap a [child] `TextField`  and
  /// automatically resizes to fit the [child]'s text value
  ///
  /// ```dart
  /// FittedTextFieldContainer(
  ///   child: TextField(...),
  /// )
  /// ```

  /// The `TextField` to wrap
  final TextField child;

  /// The calculator provides measured values that returns a width value
  final CalculateFunction calculator;

  /// The builder provides the `child` TextField to a function that returns a
  /// widget.
  final Widget Function(BuildContext context, TextField child)? builder;

  const FittedTextFieldContainer({
    Key? key,
    required this.child,
    this.builder,
    CalculateFunction? calculator,
  })  : calculator = calculator ?? FittedTextFieldCalculator.fitVisible,
        super(key: key);

  @override
  _FittedTextFieldContainerState createState() =>
      _FittedTextFieldContainerState();
}

class _FittedTextFieldContainerState extends State<FittedTextFieldContainer> {
  late double _textFieldWidth;
  late FittedTextFieldMeasurer _measurer;

  @override
  void initState() {
    assert(widget.child.controller != null);
    widget.child.controller!.addListener(_onTextChanged);
    super.initState();
  }

  @override
  void didChangeDependencies() {
    // When style is null, it defaults to `subtitle1` of current field.
    // See: https://api.flutter.dev/flutter/material/TextField/style.html
    final textStyle =
        widget.child.style ?? Theme.of(context).textTheme.subtitle1;

    _measurer = FittedTextFieldMeasurer.create(widget.child, textStyle);
    _textFieldWidth = widget.calculator(_measurer);

    super.didChangeDependencies();
  }

  @override
  void dispose() {
    widget.child.controller!.removeListener(_onTextChanged);
    super.dispose();
  }

  void _onTextChanged() {
    final width = widget.calculator(_measurer);
    if (width != _textFieldWidth) {
      setState(() {
        _textFieldWidth = width;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: _textFieldWidth,
      child: widget.builder == null
          ? widget.child
          : widget.builder!(context, widget.child),
    );
  }
}


example/lib/main.dart
import 'dart:async';
import 'dart:math' show Random;

import 'package:fitted_text_field_container/fitted_text_field_container.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Fitted Text Field Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Fitted Text Field Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController _textEditingCtl;
  FocusNode _plainFocusNode;
  FocusNode _poundFocusNode;
  FocusNode _paddingFocusNode;
  FocusNode _aniDollarFocusNode;
  FocusNode _aniDongFocusNode;
  FocusNode _aniEuroFocusNode;
  FocusNode _aniPaddingFocusNode;
  FocusNode _aniWholePoundFocusNode;

  @override
  void initState() {
    super.initState();
    _textEditingCtl = TextEditingController();
    _plainFocusNode = FocusNode();
    _poundFocusNode = FocusNode();
    _paddingFocusNode = FocusNode();
    _aniDollarFocusNode = FocusNode();
    _aniDongFocusNode = FocusNode();
    _aniEuroFocusNode = FocusNode();
    _aniPaddingFocusNode = FocusNode();
    _aniWholePoundFocusNode = FocusNode();

    Timer(Duration(milliseconds: 2000), () {
      // Uncomment this for automatic text input
      // runInputer(20);
    });
  }

  final Random rand = Random(DateTime.now().millisecondsSinceEpoch);
  void runInputer(int i) {
    Timer(Duration(milliseconds: 50 + rand.nextInt(500)), () {
      setState(() {
        _textEditingCtl.text += String.fromCharCode(97 + rand.nextInt(25));
        String num = rand.nextInt(9).toString();
        _textEditingCtl.text += num;
        if (i > 0) {
          runInputer(i - 1);
        } else {
          runDelete(10, 2000);
        }
      });
    });
  }

  void runDelete(int end, int delay) {
    Timer(Duration(milliseconds: delay), () {
      setState(() {
        _textEditingCtl.text = _textEditingCtl.text.substring(0, end);
        if (_textEditingCtl.text.length > 0) {
          runDelete(_textEditingCtl.text.length - 1,
              ((600 / 10) * end).round() + rand.nextInt(300) + 300);
        } else {
          Timer(Duration(milliseconds: 2000), () {
            runInputer(20);
          });
        }
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: ListView(
        padding: EdgeInsets.only(left: 20, right: 20, top: 10, bottom: 50),
        children: <Widget>[
          Text(
            'Static examples',
            style: Theme.of(context).textTheme.headline4,
          ),
          Fit(
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 16),
              focusNode: _plainFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.text,
              textAlign: TextAlign.left,
              maxLines: null,
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_poundFocusNode);
              },
              decoration: InputDecoration(labelText: "Multiline textfield"),
            ),
          ),
          Fit(
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 26),
              focusNode: _poundFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.numberWithOptions(decimal: true),
              maxLines: null,
              inputFormatters: [
                FilteringTextInputFormatter.allow(RegExp(r"\d+\.?\d*")),
              ],
              textAlign: TextAlign.right,
              decoration: InputDecoration(
                prefixText: "£",
                labelText: "Prefix",
              ),
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_paddingFocusNode);
              },
            ),
          ),
          Fit(
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 26),
              focusNode: _paddingFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.text,
              textAlign: TextAlign.left,
              maxLines: null,
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_aniDollarFocusNode);
              },
              decoration: InputDecoration(
                fillColor: Colors.white,
                filled: true,
                enabledBorder: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(4),
                  borderSide: BorderSide(color: Colors.grey),
                ),
                focusedBorder: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(4),
                  borderSide: BorderSide(color: Colors.blue, width: 3),
                ),
                border: OutlineInputBorder(),
                contentPadding: EdgeInsets.symmetric(
                  vertical: 15,
                  horizontal: 20,
                ),
                prefixText: "->",
                suffixText: "<-",
                hintText: "Decorations",
              ),
            ),
          ),
          Fit(
            calculator: FittedTextFieldCalculator.fitAll,
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 26),
              focusNode: _poundFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.numberWithOptions(decimal: true),
              maxLines: null,
              inputFormatters: [
                FilteringTextInputFormatter.allow(RegExp(r"\d+\.?\d*")),
              ],
              textAlign: TextAlign.right,
              decoration: InputDecoration(
                hintText: "Custom width calculator",
              ),
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_paddingFocusNode);
              },
            ),
          ),
          Text(
            'Animated examples',
            style: Theme.of(context).textTheme.headline4,
          ),
          AnimFit(
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 16),
              focusNode: _aniDollarFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.numberWithOptions(decimal: true),
              inputFormatters: [
                FilteringTextInputFormatter.allow(RegExp(r"\d+\.?\d*")),
              ],
              textAlign: TextAlign.right,
              decoration: InputDecoration(
                prefixText: "\$",
              ),
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_aniEuroFocusNode);
              },
            ),
          ),
          AnimFit(
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 26),
              focusNode: _aniEuroFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.numberWithOptions(decimal: true),
              inputFormatters: [
                FilteringTextInputFormatter.allow(RegExp(r"\d+\.?\d*")),
              ],
              textAlign: TextAlign.right,
              decoration: InputDecoration(
                suffixText: "€",
              ),
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_aniDongFocusNode);
              },
            ),
          ),
          AnimFit(
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 26),
              focusNode: _aniDongFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.number,
              inputFormatters: [
                FilteringTextInputFormatter.allow(RegExp(r"\d+")),
              ],
              textAlign: TextAlign.right,
              decoration: InputDecoration(
                suffixText: " ₫",
                hintText: "0",
              ),
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_aniPaddingFocusNode);
              },
            ),
          ),
          AnimFit(
            child: TextField(
              controller: _textEditingCtl,
              style: TextStyle(fontSize: 26),
              focusNode: _aniPaddingFocusNode,
              textInputAction: TextInputAction.next,
              keyboardType: TextInputType.number,
              inputFormatters: [
                FilteringTextInputFormatter.allow(RegExp(r"\d+")),
              ],
              textAlign: TextAlign.right,
              decoration: InputDecoration(
                fillColor: Colors.white,
                filled: true,
                enabledBorder: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(4),
                  borderSide: BorderSide(color: Colors.grey),
                ),
                focusedBorder: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(4),
                  borderSide: BorderSide(color: Colors.blue, width: 3),
                ),
                border: OutlineInputBorder(),
                contentPadding: EdgeInsets.symmetric(
                  vertical: 15,
                  horizontal: 20,
                ),
              ),
              onSubmitted: (String value) {
                FocusScope.of(context).requestFocus(_aniWholePoundFocusNode);
              },
            ),
          ),
          Text(
            'Custom builder example',
            style: Theme.of(context).textTheme.headline4,
          ),
          Center(
            child: Padding(
              padding: EdgeInsets.symmetric(vertical: 10),
              child: AnimatedFittedTextFieldContainer(
                growDuration: Duration(milliseconds: 300),
                shrinkDuration: Duration(milliseconds: 600),
                growCurve: Curves.easeOutCirc,
                shrinkCurve: Curves.easeInCirc,
                child: TextField(
                  controller: _textEditingCtl,
                  style: TextStyle(fontSize: 26),
                  focusNode: _aniWholePoundFocusNode,
                  textInputAction: TextInputAction.next,
                  keyboardType: TextInputType.number,
                  inputFormatters: [
                    FilteringTextInputFormatter.allow(RegExp(r"\d+")),
                  ],
                  textAlign: TextAlign.right,
                  decoration: InputDecoration(
                    prefixText: "£",
                    suffixText: ".00",
                    hintText: "0",
                  ),
                  onChanged: (value) => setState(() {}),
                  onSubmitted: (String value) {
                    FocusScope.of(context).requestFocus(_plainFocusNode);
                  },
                ),
                builder: (context, child) => Container(
                  child: Stack(
                    clipBehavior: Clip.none,
                    children: <Widget>[
                      child,
                      Positioned(
                        top: 0,
                        right: -22,
                        child: _textEditingCtl.text.length < 5
                            ? Icon(Icons.star_border)
                            : Icon(Icons.star, color: Colors.amber),
                      ),
                    ],
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class Fit extends StatelessWidget {
  final TextField child;
  final CalculateFunction calculator;

  Fit({Key key, @required this.child, this.calculator}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Padding(
        padding: EdgeInsets.symmetric(vertical: 10),
        child: FittedTextFieldContainer(child: child, calculator: calculator),
      ),
    );
  }
}

class AnimFit extends StatelessWidget {
  final TextField child;

  AnimFit({Key key, this.child}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Padding(
        padding: EdgeInsets.symmetric(vertical: 10),
        child: AnimatedFittedTextFieldContainer(
          growDuration: Duration(milliseconds: 300),
          shrinkDuration: Duration(milliseconds: 600),
          growCurve: Curves.easeOutCirc,
          shrinkCurve: Curves.easeInCirc,
          child: child,
        ),
      ),
    );
  }
}
